### PR TITLE
ALPN upgrade to http/2.0

### PR DESCRIPTION
**Pull request by mandarjog**
Upgrade (m)TLS connection to H2 using ALPN.

This was either a regression or perhaps never working before.

### CODE REVIEW

**Comment by mandarjog** [comment]
Current logs, note http1 connection pool only.
```diff
--- aa-master.log	2018-05-15 14:56:28.000000000 -0700
+++ aa-alpn.log	2018-05-15 14:56:05.000000000 -0700
@@ -1,5 +1,6 @@
-
 main] external/envoy/source/server/connection_handler_impl.cc:196]    new connection
+connection] external/envoy/source/common/network/connection_impl.cc:385]    socket event: 2
+connection] external/envoy/source/common/network/connection_impl.cc:453]    write ready
 connection] external/envoy/source/common/network/connection_impl.cc:385]    socket event: 3
 connection] external/envoy/source/common/network/connection_impl.cc:453]    write ready
 connection] external/envoy/source/common/network/connection_impl.cc:423]    read ready
@@ -20,8 +21,8 @@
 http] external/envoy/source/common/http/conn_manager_impl.cc:457]      'accept':'*/*'
 http] external/envoy/source/common/http/conn_manager_impl.cc:457]      ':path':'/abc'
 http] external/envoy/source/common/http/conn_manager_impl.cc:457]      ':method':'GET'
-http] external/envoy/source/common/http/conn_manager_impl.cc:670]    decode headers called: filter=0x36c4d00 status=0
-http] external/envoy/source/common/http/conn_manager_impl.cc:670]    decode headers called: filter=0x3b52980 status=0
+http] external/envoy/source/common/http/conn_manager_impl.cc:670]    decode headers called: filter=0x243b7c0 status=0
+http] external/envoy/source/common/http/conn_manager_impl.cc:670]    decode headers called: filter=0x36b7f40 status=0
 router] external/envoy/source/common/router/router.cc:250]    cluster 'outbound|80||b.istio-system.svc.cluster.local' match for URL '/abc'
 router] external/envoy/source/common/router/router.cc:298]      ':authority':'b'
 router] external/envoy/source/common/router/router.cc:298]      'user-agent':'curl/7.47.0'
@@ -29,19 +30,33 @@
 router] external/envoy/source/common/router/router.cc:298]      ':path':'/abc'
 router] external/envoy/source/common/router/router.cc:298]      ':method':'GET'
 router] external/envoy/source/common/router/router.cc:298]      'x-forwarded-proto':'http'
-router] external/envoy/source/common/router/router.cc:298]      'x-request-id':'f6687e6c-3b67-9dc7-8dd0-ae1f49afba36'
+router] external/envoy/source/common/router/router.cc:298]      'x-request-id':'43c29898-fc98-9de2-8725-256a10a7f1b2'
 router] external/envoy/source/common/router/router.cc:298]      'x-envoy-decorator-operation':'default-route'
-router] external/envoy/source/common/router/router.cc:298]      'x-b3-traceid':'d5227cd546e82734'
-router] external/envoy/source/common/router/router.cc:298]      'x-b3-spanid':'d5227cd546e82734'
+router] external/envoy/source/common/router/router.cc:298]      'x-b3-traceid':'4c7a542887eea97f'
+router] external/envoy/source/common/router/router.cc:298]      'x-b3-spanid':'4c7a542887eea97f'
 router] external/envoy/source/common/router/router.cc:298]      'x-b3-sampled':'1'
 router] external/envoy/source/common/router/router.cc:298]      'x-envoy-expected-rq-timeout-ms':'15000'
 router] external/envoy/source/common/router/router.cc:298]      ':scheme':'https'
-pool] external/envoy/source/common/http/http1/conn_pool.cc:79] creating a new connection
 client] external/envoy/source/common/http/codec_client.cc:25]    connecting
-connection] external/envoy/source/common/network/connection_impl.cc:565]    connecting to 10.48.2.38:80
+connection] external/envoy/source/common/network/connection_impl.cc:565]    connecting to 10.48.1.37:80
 connection] external/envoy/source/common/network/connection_impl.cc:574]    connection in progress
-pool] external/envoy/source/common/http/http1/conn_pool.cc:105] queueing request due to no available connections
-http] external/envoy/source/common/http/conn_manager_impl.cc:670]    decode headers called: filter=0x3b529c0 status=1
+http2] external/envoy/source/common/http/http2/codec_impl.cc:628]    setting max concurrent streams to 1073741824
+http2] external/envoy/source/common/http/http2/codec_impl.cc:635]    setting stream-level initial window size to 268435456
+http2] external/envoy/source/common/http/http2/codec_impl.cc:657]    updating connection-level initial window size to 268435456
+pool] external/envoy/source/common/http/http2/conn_pool.cc:97]    creating stream
+router] external/envoy/source/common/router/router.cc:968]    pool ready
+http2] external/envoy/source/common/http/http2/codec_impl.cc:506]    send data: bytes=24
+connection] external/envoy/source/common/network/connection_impl.cc:322]    writing 24 bytes, end_stream false
+http2] external/envoy/source/common/http/http2/codec_impl.cc:506]    send data: bytes=27
+connection] external/envoy/source/common/network/connection_impl.cc:322]    writing 27 bytes, end_stream false
+http2] external/envoy/source/common/http/http2/codec_impl.cc:460]    sent frame type=4
+http2] external/envoy/source/common/http/http2/codec_impl.cc:506]    send data: bytes=13
+connection] external/envoy/source/common/network/connection_impl.cc:322]    writing 13 bytes, end_stream false
+http2] external/envoy/source/common/http/http2/codec_impl.cc:460]    sent frame type=8
+http2] external/envoy/source/common/http/http2/codec_impl.cc:506]    send data: bytes=211
+connection] external/envoy/source/common/network/connection_impl.cc:322]    writing 211 bytes, end_stream false
+http2] external/envoy/source/common/http/http2/codec_impl.cc:460]    sent frame type=1
+http] external/envoy/source/common/http/conn_manager_impl.cc:670]    decode headers called: filter=0x36b7f80 status=1
 http] external/envoy/source/common/http/http1/codec_impl.cc:322]    parsed 68 bytes
 connection] external/envoy/source/common/network/connection_impl.cc:228]    readDisable: enabled=true disable=true
 connection] external/envoy/source/common/network/connection_impl.cc:385]    socket event: 2
@@ -59,39 +74,47 @@
 connection] external/envoy/source/common/network/connection_impl.cc:453]    write ready
 connection] external/envoy/source/common/ssl/ssl_socket.cc:99]    handshake complete
 client] external/envoy/source/common/http/codec_client.cc:63]    connected
-pool] external/envoy/source/common/http/http1/conn_pool.cc:249]    attaching to next request
-router] external/envoy/source/common/router/router.cc:968]    pool ready
-connection] external/envoy/source/common/network/connection_impl.cc:322]    writing 327 bytes, end_stream false
 connection] external/envoy/source/common/network/connection_impl.cc:453]    write ready
-connection] external/envoy/source/common/ssl/ssl_socket.cc:172]    ssl write returns: 327
+connection] external/envoy/source/common/ssl/ssl_socket.cc:172]    ssl write returns: 275
 connection] external/envoy/source/common/network/connection_impl.cc:423]    read ready
 connection] external/envoy/source/common/ssl/ssl_socket.cc:57]    ssl read returns: -1
+connection] external/envoy/source/common/network/connection_impl.cc:385]    socket event: 3
+connection] external/envoy/source/common/network/connection_impl.cc:453]    write ready
+connection] external/envoy/source/common/network/connection_impl.cc:423]    read ready
+connection] external/envoy/source/common/ssl/ssl_socket.cc:57]    ssl read returns: 37
+connection] external/envoy/source/common/ssl/ssl_socket.cc:57]    ssl read returns: -1
+http2] external/envoy/source/common/http/http2/codec_impl.cc:291]    dispatching 37 bytes
+http2] external/envoy/source/common/http/http2/codec_impl.cc:349]    recv frame type=4
+http2] external/envoy/source/common/http/http2/codec_impl.cc:349]    recv frame type=4
+http2] external/envoy/source/common/http/http2/codec_impl.cc:349]    recv frame type=8
+http2] external/envoy/source/common/http/http2/codec_impl.cc:306]    dispatched 37 bytes
+http2] external/envoy/source/common/http/http2/codec_impl.cc:506]    send data: bytes=9
+connection] external/envoy/source/common/network/connection_impl.cc:322]    writing 9 bytes, end_stream false
+http2] external/envoy/source/common/http/http2/codec_impl.cc:460]    sent frame type=4
 connection] external/envoy/source/common/network/connection_impl.cc:385]    socket event: 2
 connection] external/envoy/source/common/network/connection_impl.cc:453]    write ready
+connection] external/envoy/source/common/ssl/ssl_socket.cc:172]    ssl write returns: 9
 connection] external/envoy/source/common/network/connection_impl.cc:385]    socket event: 3
 connection] external/envoy/source/common/network/connection_impl.cc:453]    write ready
 connection] external/envoy/source/common/network/connection_impl.cc:423]    read ready
-connection] external/envoy/source/common/ssl/ssl_socket.cc:57]    ssl read returns: 541
+connection] external/envoy/source/common/ssl/ssl_socket.cc:57]    ssl read returns: 483
 connection] external/envoy/source/common/ssl/ssl_socket.cc:57]    ssl read returns: -1
-http] external/envoy/source/common/http/http1/codec_impl.cc:305]    parsing 541 bytes
-http] external/envoy/source/common/http/http1/codec_impl.cc:292]    completed header: key=date value=Tue, 15 May 2018 20:43:23 GMT
-http] external/envoy/source/common/http/http1/codec_impl.cc:292]    completed header: key=content-length value=374
-http] external/envoy/source/common/http/http1/codec_impl.cc:292]    completed header: key=content-type value=text/plain; charset=utf-8
-http] external/envoy/source/common/http/http1/codec_impl.cc:292]    completed header: key=x-envoy-upstream-service-time value=1
-http] external/envoy/source/common/http/http1/codec_impl.cc:361]    headers complete
-http] external/envoy/source/common/http/http1/codec_impl.cc:292]    completed header: key=server value=envoy
+http2] external/envoy/source/common/http/http2/codec_impl.cc:291]    dispatching 483 bytes
+http2] external/envoy/source/common/http/http2/codec_impl.cc:349]    recv frame type=1
 router] external/envoy/source/common/router/router.cc:590]    upstream headers complete: end_stream=false
-http] external/envoy/source/common/http/conn_manager_impl.cc:891]    encode headers called: filter=0x3b52940 status=0
+http] external/envoy/source/common/http/conn_manager_impl.cc:891]    encode headers called: filter=0x243b840 status=0
 http] external/envoy/source/common/http/conn_manager_impl.cc:972]    encoding headers via codec (end_stream=false):
-http] external/envoy/source/common/http/conn_manager_impl.cc:976]      'date':'Tue, 15 May 2018 20:43:23 GMT'
+http] external/envoy/source/common/http/conn_manager_impl.cc:976]      ':status':'200'
+http] external/envoy/source/common/http/conn_manager_impl.cc:976]      'date':'Tue, 15 May 2018 21:23:26 GMT'
 http] external/envoy/source/common/http/conn_manager_impl.cc:976]      'content-length':'374'
 http] external/envoy/source/common/http/conn_manager_impl.cc:976]      'content-type':'text/plain; charset=utf-8'
-http] external/envoy/source/common/http/conn_manager_impl.cc:976]      'x-envoy-upstream-service-time':'29'
+http] external/envoy/source/common/http/conn_manager_impl.cc:976]      'x-envoy-upstream-service-time':'28'
 http] external/envoy/source/common/http/conn_manager_impl.cc:976]      'server':'envoy'
-http] external/envoy/source/common/http/conn_manager_impl.cc:976]      ':status':'200'
 connection] external/envoy/source/common/network/connection_impl.cc:322]    writing 168 bytes, end_stream false
-http] external/envoy/source/common/http/conn_manager_impl.cc:1028]    encode data called: filter=0x3b52940 status=0
+http2] external/envoy/source/common/http/http2/codec_impl.cc:349]    recv frame type=0
+http] external/envoy/source/common/http/conn_manager_impl.cc:1028]    encode data called: filter=0x243b840 status=0
 http] external/envoy/source/common/http/conn_manager_impl.cc:1035]    encoding data via codec (size=374 end_stream=false)
 connection] external/envoy/source/common/network/connection_impl.cc:322]    writing 374 bytes, end_stream false
+http2] external/envoy/source/common/http/http2/codec_impl.cc:349]    recv frame type=0
 client] external/envoy/source/common/http/codec_client.cc:94]    response complete
```

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -388,20 +392,31 @@ func applyUpstreamTLSSettings(cluster *v2.Cluster, tls *networking.TLSSettings)
...
+				AlpnProtocols:     ClusterALPNProtocols,
 			},
 			Sni: tls.Sni,
 		}
 	}
+
+	if tls.Mode != networking.TLSSettings_DISABLE {
+		// alpn is set for TLS, so specify h2 configuration.
+		addHTTP2Options(cluster)
+	}
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Compare with istio mTLS. Not any mTLS.

Comment (mandarjog) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: All TLS support alpn, why not all tls?
If the other side is mtls http1.1, they would negotiate that.

Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Even the old ones?

Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: I mean old servers ?

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -339,6 +339,9 @@ func applyLoadBalancer(cluster *v2.Cluster, lb *networking.LoadBalancerSettings)
 	// DO not do if else here. since lb.GetSimple returns a enum value (not pointer).
 }
 
+// ClusterALPNProtocols - try h2 first and then http/1.1
+var ClusterALPNProtocols = []string{"h2", "http/1.1"}
```
[the remark this answers is not in the document]
Comment (mandarjog) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: @PiotrSikora this code is not envoy specific. 
Per alpn spec client should be able to claim what protocols it supports, right?

[the remark this answers is not in the document]
Comment (PiotrSikora) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: But the code is wrong when used with Envoy, i.e. you'll announce `h2,http/1.1`, but never use HTTP/1.1, even if negotiated.

**Comment by mandarjog** [comment]
@PiotrSikora, h2 is now upgraded only in the mesh.
For out of mesh, if the cluster is already h2, we specify `{"h2"}` in alpn.

**Pull request merged**